tests/Struct/FindologicProductTest.php

```diff
@@ -189,6 +188,86 @@ public function testNoProductCategories(): void
         );
     }
 
+    public function parentAndChildrenCategoryProvider(): array
+    {
+        return [
+            'Parent and children have the same categories assigned' => [
+                'isParentAssigned' => true,
+                'isVariantAssigned' => true,
+            ],
+            'Parent has no categories and children have some categories assigned' => [
+                'isParentAssigned' => false,
+                'isVariantAssigned' => true
+            ],
+            'Parent has categories and children have no categories assigned' => [
+                'isParentAssigned' => true,
+                'isVariantAssigned' => false
+            ],
+        ];
```

Comment:
A data provider should hold `data`, as the name suggests. Here you are simply holding `boolean` values to change the behavior of the test. This is not the purpose of a data provider. Change this data provider to return the product data instead of boolean values, or refactor this into multiple test cases, whatever is easier and makes more sense in the context.

Reply:
The data is the same that we need to pass, so it does not make sense to clutter the provider with the same array of data. We have used providers like this in several other places too.

Reply:
Yeah it's just bad, because we can't extend this test in the future, without making changes in the test itself. Maybe we can refactor this in the future, it was just a minor point anyway.